synapse/rest/client/v2_alpha/account.py

```diff
@@ -224,38 +155,45 @@ def on_POST(self, request):
         return (200, ret)
 
 
-class PasswordResetSubmitTokenServlet(RestServlet):
+class ThreepidSubmitTokenServlet(RestServlet):
     """Handles 3PID validation token submission"""
 
     PATTERNS = client_patterns(
-        "/password_reset/(?P<medium>[^/]*)/submit_token/*$", releases=(), unstable=True
+        "/(?P<purpose>registration|password_reset)/(?P<medium>[^/]*)/submit_token/*$",
+        releases=(),
+        unstable=True,
     )
 
     def __init__(self, hs):
         """
         Args:
             hs (synapse.server.HomeServer): server
         """
-        super(PasswordResetSubmitTokenServlet, self).__init__()
+        super(ThreepidSubmitTokenServlet, self).__init__()
         self.hs = hs
         self.auth = hs.get_auth()
         self.config = hs.config
         self.clock = hs.get_clock()
         self.datastore = hs.get_datastore()
 
     @defer.inlineCallbacks
-    def on_GET(self, request, medium):
+    def on_GET(self, request, purpose, medium):
+        if purpose == "password_reset":
+            purpose = "password resets"
```

Comment:
I think overwriting the input param is a bit confusing, and it doesn't really work (you end up with strings like "User registration have been disabled" and "Email-based password resets is disabled". Better to make each message conditional as a whole.

(In fact, it might end up being clearer to have two separate servlets and factor out the common code to a base class or to utility functions, but I don't know if that would work out cleaner or not)